src/replica/mutation_log.cpp

```diff
@@ -916,6 +916,62 @@ int64_t mutation_log::total_size_no_lock() const
     return _log_files.size() > 0 ? _global_end_offset - _global_start_offset : 0;
 }
 
+error_code mutation_log::reset_from(const std::string &dir,
+                                    replay_callback replay_error_callback,
+                                    io_failure_callback write_error_callback)
+{
+    error_code err = ERR_FILE_OPERATION_FAILED;
+
+    // close for flushing current log and be ready to open new log files after reset
+    close();
+
+    // make sure logs in `dir` (such as /learn) are valid.
+    error_s es = log_utils::check_log_files_continuity(dir);
+    if (!es.is_ok()) {
+        derror_f("the log of source dir {} is valid:{}, will remov it.", dir, es);
+        if (!utils::filesystem::remove_path(dir)) {
+            derror_f("remove {} failed", dir);
+            return err;
+        }
+        return es.code();
+    }
+
+    std::string temp_dir = _dir + '.' + std::to_string(dsn_now_ns());
+    if (!utils::filesystem::rename_path(_dir, temp_dir)) {
+        derror_f("rename {} to {} failed", _dir, temp_dir);
+        return err;
+    }
+    ddebug_f("moved current log dir {}  to tmp_dir {}", _dir, temp_dir);
+
+    // move source dir to target dir
+    if (!utils::filesystem::rename_path(dir, _dir)) {
+        derror_f("rename {} to {} failed", dir, _dir);
+        return err;
+    }
+    ddebug_f("move {} to {} as our new log directory", dir, _dir);
+
+    // - make sure logs in moved dir(such as /plog) are valid and can be opened successfully.
+    // - re-open new log files  for loading the new log file and register the files into replica,
+    // please make sure the old log files has been closed
+    err = open(replay_error_callback, write_error_callback);
+    if (err != ERR_OK) {
+        derror_f(
+            "the logs of moved dir {} are valid and open failed:{}, will rollback it.", _dir, err);
```

Comment:
What is this error log meaning? the logs are valid but open failed?

Reply:
also should `invalid`